lib/phoenix/endpoint.ex

```diff
@@ -398,8 +398,16 @@ defmodule Phoenix.Endpoint do
   defp config(opts) do
     quote do
       @otp_app unquote(opts)[:otp_app] || raise "endpoint expects :otp_app to be given"
+
       var!(config) = Phoenix.Endpoint.Supervisor.config(@otp_app, __MODULE__)
-      var!(code_reloading?) = var!(config)[:code_reloader]
+
+      # Compile-time configuration checkg
+      #
+      # This ensures that if a compile-time configuration is overwritten at runtime the application won't boot.
+      var!(code_reloading?) = Application.compile_env(@otp_app, [__MODULE__, :code_reloader], var!(config)[:code_reloader])
+      var!(debug_errors?) =  Application.compile_env(@otp_app, [__MODULE__, :debug_errors], var!(config)[:debug_errors])
+      var!(render_errors) =  Application.compile_env(@otp_app, [__MODULE__, :render_errors], var!(config)[:render_errors])
```

Comment:
```suggestion
      var!(code_reloading?) = Application.compile_env(@otp_app, [__MODULE__, :code_reloader])
      var!(debug_errors?) =  Application.compile_env(@otp_app, [__MODULE__, :debug_errors])
      var!(force_ssl) =  Application.compile_env(@otp_app, [__MODULE__, :force_ssl])
```

Reply:
I was going to suggest the removal of the checks done at the Endpoint supervisor but I see you already did that! =)